portality/bll/services/article.py

```diff
@@ -0,0 +1,375 @@
+
+            ft = article.get_normalised_fulltext()
+            if ft in fulltexts:
+                return True
+            fulltexts.append(ft)
+
+        return False
+
+
+    def create_article(self, article, account, duplicate_check=True, merge_duplicate=True, limit_to_account=True, dry_run=False):
+        """
+        Create an individual article in the database
+
+        This method will check and merge any duplicates, and report back on successes and failures in a manner consistent with
+        batch_create_articles.
+
+        :param article: The article to be created
+        :param account:     The account creating the article
+        :param duplicate_check:     Whether to check for duplicates in the database
+        :param merge_duplicate:     Whether to merge duplicate if found.  If set to False, may result in a DuplicateArticleException
+        :param limit_to_account:    Whether to limit create to when the account owns the journal to which the article belongs
+        :param dry_run:     Whether to actuall save, or if this is just to either see if it would work, or to prep for a batch ingest
+        :return:
+        """
+        # first validate the incoming arguments to ensure that we've got the right thing
+        argvalidate("create_article", [
+            {"arg": article, "instance" : models.Article, "allow_none" : False, "arg_name" : "article"},
+            {"arg": account, "instance" : models.Account, "allow_none" : False, "arg_name" : "account"},
+            {"arg" : duplicate_check, "instance" : bool, "allow_none" : False, "arg_name" : "duplicate_check"},
+            {"arg" : merge_duplicate, "instance" : bool, "allow_none" : False, "arg_name" : "merge_duplicate"},
+            {"arg" : limit_to_account, "instance" : bool, "allow_none" : False, "arg_name" : "limit_to_account"},
+            {"arg" : dry_run, "instance" : bool, "allow_none" : False, "arg_name" : "dry_run"}
+        ], exceptions.ArgumentException)
+
+        if limit_to_account:
+            legit = self.is_legitimate_owner(article, account.id)
+            if not legit:
+                owned, shared, unowned, unmatched = self.issn_ownership_status(article, account.id)
+                return {"success" : 0, "fail" : 1, "update" : 0, "new" : 0, "shared" : shared, "unowned" : unowned, "unmatched" : unmatched}
+
+        # before saving, we need to determine whether this is a new article
+        # or an update
+        is_update = 0
+        if duplicate_check:
+            duplicate = self.get_duplicate(article, account.id)
+            if duplicate is not None:
+                if merge_duplicate:
+                    is_update  = 1
+                    article.merge(duplicate) # merge will take the old id, so this will overwrite
+                else:
+                    raise exceptions.DuplicateArticleException()
+
+        # finally, save the new article
+        if not dry_run:
+            article.save()
```

Comment:
can the save() method throw up any surprises leading to our success response being a lie?

Reply:
The new version of the save method does some fairly aggressive error handling and reporting, so if save fails, even after several re-tries, we would expect an exception.  I don't think there's any further handling that we need to do here - the exception would cause the front end to get an error, but that's probably what we want.